### DISCUSSION TITLE
pkgx stderr output interfering with neovim plugins

### DISCUSSION THREAD

**Discussion by fredrikaverpil** [pkgxdev/pkgx#916] [Help]
I just got hit by [an issue](https://github.com/nvim-neotest/neotest-go/pull/54#issuecomment-1880791338) ~~seemingly stemming from that Go is compiled in a way so that it causes printouts of compiler warnings~~.

~~This is not how Go provided by Homebrew behaves. Would it make sense to try and sync the way you build Go to the way Homebrew builds Go?~~

EDIT: I was mistaken, title updated to better match actual problem.

**Comment by michaelessiet** [comment]
@jhheider what are your thoughts on this. Is this something that we can fix or is it out of our control?

**Reply by jhheider** [reply]
this doesn't appear to be specifically go related, unless i'm misreading it, but rather output from `pkgx`, which should all be going to stdout. @fredrikaverpil can you confirm that you only get the `pkgx` output (resolving graph, etc) on first invocation? it should be otherwise silent on subsequent invocations, i believe.

**Reply by fredrikaverpil** [reply]
@jhheider oh shoot!
I totally did not realize this `resolving graph…` is of course the thing you get directly from `pkgx`... 🤔 

So `pkgx go@latest version` always shows the "resolving graph ..." each time I run that command. But it doesn't look like running `go version` (provided by `pkgx install go@latest`) does it.

But in my case, I'm using a Neovim plugin (neotest-go) which invokes `go` ([here](https://github.com/nvim-neotest/neotest-go/blob/d29d20d912aca81a07c50022d880cc66f0d26542/lua/neotest-go/init.lua#L172)). It looks like it always manages to hit the "resolving graph..." on every invocation.

**Reply by jhheider** [reply]
consider using `pkgx install go.dev` which will add a shim to `~/.local/bin/go`. i _think_ that resolves the issue:
```shell
$ pkgx install go.dev

pkgx: installed: ~/.local/bin/go
pkgx: installed: ~/.local/bin/gofmt

$ go 2>foo
$ cat foo
Go is a tool for managing Go source code.

Usage:

...
```

**Reply by fredrikaverpil** [reply]
@jhheider this did not make a difference for me. I'm still seeing the "resolving graph..." string crashing the neotest-go plugin.

**Reply by jhheider** [reply]
hm, what if you add the `--silent` flag to the `~/.local/bin/go` shim?
```sh
#!/bin/sh
if [ "$PKGX_UNINSTALL" != 1 ]; then
  exec pkgx --silent +go.dev -- go "$@"
else
  cd "$(dirname "$0")"
  rm go gofmt && echo "uninstalled: go.dev" >&2
fi
```

**Reply by fredrikaverpil** [reply]
Yes, adding the `--silent` in fixed the crash, actually!
Is this something we can add to the go.dev shim?

**Reply by jhheider** [reply]
possibly, that should be a default for shims for this reason.

@mxcl what do you think?

**Reply by mxcl** [reply]
I think I don't understand the problem. The question is about compiling go and the solution is to silence the shim…?

**Reply by mxcl** [reply]
The resolving message goes to stderr *AND* only prints if stderr is a TTY. So what is neovim doing exactly that makes us think it's a TTY? And why are they reading stderr for stdout? Do they not respect UNIX conventions? Are they the “bad ones”?

**Reply by jhheider** [reply]
The resolving output of pkgx was messing up the neovim plugin. Silencing the shim lets it behave like "normal" go. Was wondering if you think silencing shims in general might be wise. Or lowering their verbosity.

**Reply by IpongJkBase** [reply]
Yess

**Reply by Lazycryp** [reply]
that's it.....

**Reply by mxcl** [reply]
K I replied in that ticket to get more info. As said we already take two precautions to not output unless the recipient really wants it. This is the first time we have found a project where that is not enough.

I don't want to silence all output for shims. That would mean that a completely undefined amount of time may elapse with no information to the user at all about what is happening if projects need to be cached.

So lets see what this project says. I would be interested to understand why they are reading stderr and why they are presenting themselves to us as a terminal.

**Reply by mxcl** [reply]
In the meantime, edit the `go` shim and add `--silent` after `pkgx`. Once we understand what's going on here we'll try to do what is best for everyone.

**Comment by sambojix** [comment]
I think this is a natural thing, because this can be an input for tea developers to look for mistakes to build a more advanced ecosystem in the future....This can be an input....and it is natural because this is just the beginning🙏🙏

**Reply by HubbusBubbus** [reply]
It's fine
